content/courses/Ada_For_The_C_Embedded_Developer/chapters/08_Performance.rst

```diff
@@ -240,6 +240,124 @@ performance-critical pieces of the application.
 Pointers v.s. data copies
 -------------------------
 
-.. todo::
+In the section about :ref:`pointers <Pointers>`, we mentioned that the Ada
+compiler will automatically pass parameters by reference when needed. Also, in
+the section about :ref:`by value vs. by reference <By_Value_Vs_By_Reference>`,
+we mentioned that scalar types and pointers are passed by value, while record
+and array types are always passed by reference. Therefore, unlike C, you don't
+have to use access types in Ada to get better performance when passing arrays
+or records to subprograms. Using :ada:`out` or :ada:`in out` gives you
+equivalent performance and, at the same time, improves the readability of your
+code.
 
-    Complete section!
+Let's look at this example:
+
+[C]
+
+.. code:: c manual_chop run_button project=Courses.Ada_For_C_Embedded_Dev.Performance.Passing_Rec_By_Reference_C
+
+    !main.c
+    #include <stdio.h>
+
+    struct Data {
+        int prev, curr;
+    };
+
+    void update(struct Data *d,
+                int          v)
+    {
+        d->prev = d->curr;
+        d->curr = v;
+    }
+
+    void display(const struct Data *d)
+    {
+        printf("Prev : %d\n", d->prev);
+        printf("Curr : %d\n", d->curr);
+    }
+
+    int main(int argc, const char * argv[])
+    {
+        struct Data D1 = { 0, 1 };
+
+        update (&D1, 3);
+        display (&D1);
+    }
+
+In this C code example, we're using pointers to pass :c:`D1` as a reference to
+:c:`update` and :c:`display`. In contrast, the equivalent code in Ada makes use
+of parameter modes to pass :ada:`D1` as an argument to :ada:`Update` and
```

Comment:
The modes don't affect the compiler's parameter-passing mechanism selection, so maybe "In contrast, the equivalent code in Ada simply uses the parameter modes to specify the data flow directions. The mechanisms used to pass the values do not appear in the source code."

Reply:
Thanks. I've just updated the description.